csharp/src/AddressBook/Addressbook.pb.cs

```diff
@@ -55,34 +77,37 @@ public sealed partial class Person : pb::IMessage<Person>
       , pb::IBufferMessage
   #endif
   {
+    private static readonly bool _runtimeVersionValidated = global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.ValidateRuntimeVersion();
+
     private static readonly pb::MessageParser<Person> _parser = new pb::MessageParser<Person>(() => new Person());
     private pb::UnknownFieldSet _unknownFields;
     [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
-    [global::System.CodeDom.Compiler.GeneratedCode("protoc", null)]
+    [global::System.CodeDom.Compiler.GeneratedCode("protoc", global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.ConstProtocVersion)]
     public static pb::MessageParser<Person> Parser { get { return _parser; } }
 
     [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
-    [global::System.CodeDom.Compiler.GeneratedCode("protoc", null)]
+    [global::System.CodeDom.Compiler.GeneratedCode("protoc", global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.ConstProtocVersion)]
     public static pbr::MessageDescriptor Descriptor {
       get { return global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.Descriptor.MessageTypes[0]; }
     }
 
     [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
-    [global::System.CodeDom.Compiler.GeneratedCode("protoc", null)]
+    [global::System.CodeDom.Compiler.GeneratedCode("protoc", global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.ConstProtocVersion)]
     pbr::MessageDescriptor pb::IMessage.Descriptor {
       get { return Descriptor; }
     }
 
     [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
-    [global::System.CodeDom.Compiler.GeneratedCode("protoc", null)]
+    [global::System.CodeDom.Compiler.GeneratedCode("protoc", global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.ConstProtocVersion)]
     public Person() {
+      global::Google.Protobuf.Examples.AddressBook.AddressbookReflection.ValidateRuntimeVersion();
```

Comment:
Why do we need this in the constructor. Wouldn't the fact that we are initializing the static field with this call make this redundant?

Reply:
Nope - unless someone "touches" a static field (e.g. _parser) the static field initializer doesn't have to be run - and in my tests, it *doesn't* get run, in at least some environments (e.g. net6.0). Sample code:

```csharp
using System;

class Program
{
    static void Main()
    {
        Test t = new Test();
    }
}

class Test
{
    static string x = Log("This is from an initializer");

    static string Log(string x)
    {
        Console.WriteLine(x);
        return x;
    }

    public Test()
    {
        Console.WriteLine("In constructor");
    }
}
```

Reply:
(Whereas the timing requirements for a static constructor are far more detailed, and *would* include creating an instance.)

Reply:
Ohh that's a shame, well, not really, that's good for all other cases.

But then, what is the purpose of the static field? It won't be initialized unless it is touched, and no one is touching it.
And that also means that if only static members of a message class are used, then the check is not run for that message?
Which means that we need to go back and rely on the static constructor of each message, which is the only thing that executes whenever any member of a class is touched.

Or am I missing something again?

Reply:
And another question, is this only taking care of messages? What about services? Although unlikely I guess, there could only be parameterless void services (or with primitive types) in a given proto file, no?

Reply:
Purpose of static field: it allows problems to be spotted before access to the parser, for example. We could *potentially* drop that, at the cost of possibly having some less clear stack traces.

Services: we don't actually generate anything from proto services in protoc. That's all done by the gRPC generator.

Reply:
> it allows problems to be spotted before access to the parser, for example. We could potentially drop that, at the cost of possibly having some less clear stack traces.

Because if one static field is touched (i.e. the parser) all static fields are initialized?